File: examples/haskell/pink-fork-freebsd.hs
-- A simple example showing how to do the tracing fork on FreeBSD
-- vim: set ft=haskell et ts=4 sts=4 sw=4 fdm=marker :

import Data.Maybe
import Control.Monad
import System.Exit
import System.Posix.Process
import System.Posix.Signals
import System.PinkTrace.Trace

main :: IO ()
main = do
    pid <- forkProcess child
    status <- getProcessStatus True False pid
    case fromJust status of
        Exited (ExitFailure ret) -> putStrLn ("Child exited with code: " ++ (show ret)) >> exitFailure
        Exited ExitSuccess       -> putStrLn ("Child exited with code: 0") >> exitFailure
        Terminated sig           -> putStrLn ("Child was terminated: " ++ (show sig)) >> exitFailure
        Stopped sig              -> unless (sig == softwareStop)
                                        (putStrLn ("Wrong signal: " ++ show sig) >> exitFailure)

    -- Nothing interesting in this example.
    -- Let the child continue its execution.
    traceContinue 1 nullSignal pid
    where
        child :: IO ()
        child = do
            traceMe
            raiseSignal softwareStop
            putStrLn "hello world"

File: examples/c/pink-fork-freebsd.c
/**
 * \file
 *
 * Example \ref pink-fork-freebsd.c "pink-fork-freebsd.c" .
 **/

/**
 * \example pink-fork-freebsd.c
 *
 * Example demonstrating how to fork and start tracing a process on FreeBSD.
 **/

#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>

#include <pinktrace/pink.h>

int
main(void)
{
	int status;
	pid_t pid;

	/* Fork and start tracing. */
	if ((pid = fork()) < 0) {
		perror("fork");
		return EXIT_FAILURE;
	}
	else if (!pid) {
		/* Set up for tracing. */
		if (!pink_trace_me()) {
			perror("pink_trace_me");
			_exit(EXIT_FAILURE);
		}
		/* Stop, so that the parent has the chance to set up options
		 * for tracing. */
		kill(getpid(), SIGSTOP);

		/* Here's the point where you can do whatever the traced child
		 * should do. */
		printf("hello world\n");

		_exit(EXIT_SUCCESS);
	}
	else {
		waitpid(pid, &status, 0);
		assert(WIFSTOPPED(status));
		assert(WSTOPSIG(status) == SIGSTOP);

		/* The child is ready for tracing, nothing interesting in this
		 * example, let the child resume its execution. */
		pink_trace_resume(pid, 0);

		/* Wait for the child to exit. */
		waitpid(pid, &status, 0);

		return EXIT_SUCCESS;
	}
}
